Documentation/devel/inspect-containers.md

```diff
@@ -0,0 +1,354 @@
+```
+
+We can see that rkt mounts the stage1 and stage2 filesystems with the [tree store][treestore] as `lowerdir` on the directory rkt expects them to be.
+
+This means that, for a same tree store, everything will be shared in a copy-on-write (COW) manner, except the bits that each container modifies, which will be in the `upperdir` and will appear magically in the mount destination.
+You can read more about this filesystem in the [overlay documentation][overlay].
+
+
+### stage1
+
+This is where most of the interesting things happen, the first being executing the stage1 [run entrypoint][run-entrypoint], which is `/init` by default:
+
+```
+5710  execve("/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/stage1/rootfs/init", ["/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/stage1/rootfs/init", "--net=default", "--local-config=/etc/rkt", "d5513d49-d14f-45d1-944b-39437798ddda"], 0xc42009b040 /* 25 vars */ <unfinished ...>
+```
+
+init does a bunch of stuff, including creating the container's network namespace and mounting a reference to it on the host filesystem:
+
+```
+5723  unshare(CLONE_NEWNET)             = 0
+5723  mount("/proc/5710/task/5723/ns/net", "/var/run/netns/cni-eee014d2-8268-39cc-c176-432bbbc9e959", 0xc42017c6a8, MS_BIND, NULL) = 0
+```
+
+After creating the network namespace, it will execute the relevant [CNI][cni] plugins.
+The default network uses the [ptp plugin][ptp] with [host-local][host-local] as IPAM:
+
+```
+5725  execve("stage1/rootfs/usr/lib/rkt/plugins/net/ptp", ["stage1/rootfs/usr/lib/rkt/plugins/net/ptp"], 0xc4201ac000 /* 32 vars */ <unfinished ...>
+5730  execve("stage1/rootfs/usr/lib/rkt/plugins/net/host-local", ["stage1/rootfs/usr/lib/rkt/plugins/net/host-local"], 0xc42008e240 /* 32 vars */ <unfinished ...>
+```
+
+In this case, the CNI plugins use come from rkt's stage1, but [rkt is also able to pick a CNI plugin installed externally](https://github.com/rkt/rkt/blob/v1.29.0/Documentation/networking/overview.md#custom-plugins).
+
+The plugins will do some iptables magic to configure the network:
+
+```
+5739  execve("/usr/bin/iptables", ["iptables", "--version"], 0xc42013e000 /* 32 vars */ <unfinished ...>
+5740  execve("/usr/bin/iptables", ["/usr/bin/iptables", "-t", "nat", "-N", "CNI-7a59ad232c32bcea94ee08d5", "--wait"], 0xc4200b0a20 /* 32 vars */ <unfinished ...>
+...
+```
+
+After the network is configured, rkt mounts the container cgroups instead of letting systemd-nspawn do it because we want to have control on how they're mounted.
+We also mount the host cgroups if they're not already mounted in the way systemd-nspawn expects them.
+
+We do this in a new mount namespace to avoid polluting the host mounts and to get automatic cleanup when the container exits:
+
+```
+5710  unshare(CLONE_NEWNS)              = 0
```

Comment:
you can explain that CLONE_NEWNS is the flag for mount namespace. It is not immediately visible, but you can search the explanation of that flag in "man 2 clone".

Reply:
since this is the first kind of namespace implemented in Linux, it does not have the "mount" or MNT word in it...

Reply:
actually, http://man7.org/linux/man-pages/man7/namespaces.7.html is a more readable source of information for the flags, as you mention below..